js/app/utils.js

```diff
@@ -560,3 +561,34 @@ function intersection(a, b) {
         .forEach(item => intersection.add(item));
     return [...intersection];
 }
+
+function parse_uri (uri_str, parse_query) {
+    let uri = new SoupURI (uri_str);
+
+    if (!uri)
+        return null;
+
+    let query = uri.query;
+
+    if (parse_query && query) {
+        let params = query.split('&');
+        query = {};
+
+        params.forEach((param) => {
+            let tokens = param.split('=');
+            query[decodeURIComponent(tokens[0])] = decodeURIComponent(tokens[1]);
+        });
+    }
+
+    /* Make a JS object copy from SoupURI */
+    return {
```

Comment:
You can accomplish this shorter, though maybe not more readable, with ES6 syntax:
```js
let {scheme, user, password, host, port, path, fragment} = uri;
return {scheme, user, password, host, port, path, query, fragment};
```